Extension/src/LanguageServer/configurations.ts

```diff
@@ -241,26 +242,27 @@ export class CppProperties {
         });
 
         this.configFileWatcher.onDidChange(() => {
-            // If the file is one of the textDocument's vscode is tracking, we need to wait for an
-            // onDidChangeTextDocument event, or we may get old/cached contents when we open it.
-            let alreadyTracking: boolean = false;
-            for (let i: number = 0; i < vscode.workspace.textDocuments.length; i++) {
-                if (vscode.workspace.textDocuments[i].uri.fsPath === settingsPath) {
-                    alreadyTracking = true;
-                    break;
-                }
-            }
-            if (!alreadyTracking) {
-                this.handleConfigurationChange();
-            }
+            this.handleConfigurationChange();
         });
 
-        vscode.workspace.onDidChangeTextDocument((e: vscode.TextDocumentChangeEvent) => {
-            if (e.document.uri.fsPath === settingsPath) {
-                this.handleConfigurationChange();
+        vscode.workspace.onDidChangeTextDocument((e) => {
+            if (e.document.uri.fsPath === settingsPath && this.isCppPropertiesJsonVisible) {
+                this.handleSquiggles();
             }
         });
 
+        vscode.window.onDidChangeVisibleTextEditors((editors) => {
+            let wasVisible: boolean = false;
+            wasVisible = this.isCppPropertiesJsonVisible;
+            this.isCppPropertiesJsonVisible = false;
+            editors.forEach(editor => {
+                if (editor.document.uri.fsPath  === settingsPath && !wasVisible) {
+                    this.isCppPropertiesJsonVisible = true;
+                    this.handleSquiggles();
+                }
+            });
+        });
```

Comment:
A couple of problems remaining in this function.

```
let wasVisible: boolean = false;
wasVisible = this.isCppPropertiesJsonVisible;
```

Can be:

```
let wasVisible: boolean = this.isCppPropertiesJsonVisible;
```

And because you're only setting `isCppPropertiesJsonVisible` (back) to true if `!wasVisible`, you'll wind up leaving it as false whenever some other document becomes visible (even if cpp_properties is still also visible).

I'd suggest:

```
if (editor.document.uri.fsPath  === settingsPath) {
    this.isCppPropertiesJsonVisible = true;
    if (!wasVisible)
        this.handleSquiggles();
}
```